Navigation: Use a single placeholder for the block

tellthemachines opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

Closes #35818.

Goes back to the single-step placeholder for Navigation block. It's essentially the second step of the current 2-step flow, where:

* existing custom post type menus and existing classic menus are brought together under the same "Add existing menu" dropdown and
* "Add new pages" and "Start empty" options pop up a dialog to choose a name for the new menu.

One thing that might be confusing is that CPTs and classic menus are now indistinguishable from each other in the dropdown. It might be good to have two sections in there with some sort of heading before each. Ideas appreciated!

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

Add a Navigation block; try each of the three options that appear.

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
New feature (non-breaking change which adds functionality)
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

tellthemachines committed:
Try single placeholder for Navigation block

jasmussen commented:
Fast work! Here's the one-step process selecting an existing menu:
![onestep](https://user-images.githubusercontent.com/1204802/139221502-9deeedcc-9d31-41d8-8fe3-017101820580.gif)

Here's creating a new one:
![start empty](https://user-images.githubusercontent.com/1204802/139221525-c768cc4a-0d54-4319-bab0-258e03dd39e0.gif)

In both cases, these take big steps towards managing the weight of the placeholder as it was before, nice work. There are still some larger issues to solve, but those are unrelated to this PR.

 A few small things we should fix, though:
<img width="946" alt="Screenshot 2021-10-28 at 10 48 17" src="https://user-images.githubusercontent.com/1204802/139221751-e3ee0405-c470-4862-9b08-c98ad182e638.png">

The footprint of the block is larger than the placeholder. This appears to be because the `gap` property is spacing out the placeholder with a separate div that sits below it:

<img width="689" alt="Screenshot 2021-10-28 at 10 51 32" src="https://user-images.githubusercontent.com/1204802/139221960-5894ac8d-0168-4417-8e79-1450852df654.png">
<img width="770" alt="Screenshot 2021-10-28 at 10 51 28" src="https://user-images.githubusercontent.com/1204802/139221973-85bd4ab3-ada6-408d-9afe-f0e3cb852c1c.png">

This seems like it could be an existing issue, and this bit of CSS appears to fix it:

```
.components-placeholder.wp-block-navigation-placeholder + .wp-block-navigation__responsive-container {
    display: none;
}
```

CC: @vcanales for good measure.

I also wonder if we shouldn't revisit the decision to make blue primary buttons. This screams a bit much with two buttons:

<img width="701" alt="Screenshot 2021-10-28 at 10 55 02" src="https://user-images.githubusercontent.com/1204802/139222619-059b3811-f339-4dcd-aff6-fc405c41dd04.png">

A small fix might be to just ensure there's only one primary blue button at a time. But I think it might be time to move towards this simpler layout:
<img width="1520" alt="Start empty" src="https://user-images.githubusercontent.com/1204802/139223352-49861e68-cc26-4c45-bd21-9a37b01973f1.png">

What do you think?

That design takes small steps towards some of the ideas [here](https://github.com/WordPress/gutenberg/issues/35947#issuecomment-952738005).

jasmussen commented:
> One thing that might be confusing is that CPTs and classic menus are now indistinguishable from each other in the dropdown. It might be good to have two sections in there with some sort of heading before each. Ideas appreciated!

I forgot to respond to that one. Here's a mockup that adds separators and subheadings to the submenu:
<img width="1520" alt="dropdown" src="https://user-images.githubusercontent.com/1204802/139229621-50158263-f995-476a-af1b-97edd9986ef5.png">

I don't think we necessarily need to get all the way to that design immediately — we can think of that more as a general pointer. But I do think separating things out could make sense. 

The verbiage is also up for grabs — I've currently used "Synced" as a term as inspired by a potential renaming of [Reusable Blocks to Synced Blocks](#34352). 

We might even be fine with just calling it "Menus":

<img width="870" alt="Screenshot 2021-10-28 at 11 36 40" src="https://user-images.githubusercontent.com/1204802/139229964-9facd3c8-7d1c-43ef-a15c-c3111bfb528b.png">

tellthemachines commented:
> I also wonder if we shouldn't revisit the decision to make blue primary buttons. This screams a bit much with two buttons

Hmm, that's weird. Here's what I'm seeing:

<img width="866" alt="Screen Shot 2021-10-29 at 2 57 08 pm" src="https://user-images.githubusercontent.com/8096000/139373254-a468d1d6-ef4d-4a24-bbbe-e85588c7d22b.png">

tellthemachines committed:
Fix extra space caused by responsive container

tellthemachines committed:
Separate menus from classic menus and add labels.

tellthemachines commented:
I added a CSS fix for that extra space and updated the dropdown to use separate MenuGroups with labels for Menus and Classic Menus.

jasmussen commented:
This is fast coming together, thank you for the work. The focus style trick appears to have worked:
<img width="961" alt="Screenshot 2021-10-29 at 08 26 26" src="https://user-images.githubusercontent.com/1204802/139386277-c9cdc40f-65e6-4fd5-80e7-f227c6c3c380.png">

The submenu also has a nice subheading:
<img width="796" alt="Screenshot 2021-10-29 at 08 26 29" src="https://user-images.githubusercontent.com/1204802/139386300-fa9132e0-9b56-4ab7-bf9f-9ec796ee437d.png">

Can you make the popover "isAlternate"? That will give it a dark border and appear to be more related to the in-canvas block UI. [It's supposed to be enforced](https://github.com/WordPress/gutenberg/pull/33129), but it doesn't appear to work here.

I also still see those two blue buttons. Not sure exactly why.

jasmussen commented:
The odd positioning of the popover is tracked somewhere in the archive — it's due to the positioning math of the dropdown calculating based on the center of its opening button, rather than from the left edge.

vcanales reviewed (commented):
> The footprint of the block is larger than the placeholder. This appears to be because the gap property is spacing out the placeholder with a separate div that sits below it.

Regarding this, not rendering `ResponsiveWrapper` when the Placeholder is shown would be better than hiding it, ie. the following instead of a CSS rule:

```js
{ ! isPlaceholderShown && (
  <ResponsiveWrapper
    id={ clientId }
    onToggle={ setResponsiveMenuVisibility }
    isOpen={ isResponsiveMenuOpen }
    isResponsive={ 'never' !== overlayMenu }
    isHiddenByDefault={ 'always' === overlayMenu }
 >
    { isEntityAvailable && (
      <NavigationInnerBlocks
        isVisible={ ! isPlaceholderShown }
        clientId={ clientId }
        appender={ CustomAppender }
        hasCustomPlaceholder={
          !! CustomPlaceholder
        }
        orientation={ orientation }
      />
    ) }
  </ResponsiveWrapper>
) }
```

tellthemachines committed:
Add popover prop.

tellthemachines committed:
Don't render responsive wrapper with placeholder.

talldan commented on packages/block-library/src/navigation/edit/placeholder/index.js:
```diff
@@ -50,30 +66,194 @@ export default function Placeholder( {
+					<PlaceholderPreview />
+
+					<div className="wp-block-navigation-placeholder__controls">
+						{ isStillLoading && (
+							<div>
+								<Spinner />
+							</div>
+						) }
```
Maybe this can be consolidated with the logic on line 138, so that there's only one loading state (`<PlaceholderPreview isLoading />`).

talldan commented on packages/block-library/src/navigation/edit/placeholder/index.js:
```diff
@@ -50,30 +66,194 @@ export default function Placeholder( {
+								<Spinner />
+							</div>
+						) }
+						{ ! isStillLoading && (
```
If the loading state is consolidated, it should be possible to move this up to line 139 to reduce the nesting.

talldan commented on packages/block-library/src/navigation/edit/placeholder/index.js:
```diff
@@ -50,30 +66,194 @@ export default function Placeholder( {
+									{ __( 'Navigation' ) }
+								</div>
+								{ hasMenus || canSwitchNavigationMenu ? (
+									<DropdownMenu
```
This code gets very deeply nested. It's very hard to read a diff. I think it'd be good to consider breaking the placeholder down into separate components, or maybe just extract this dropdown menu into a separate component.

talldan commented on packages/block-library/src/navigation/edit/placeholder/index.js:
```diff
@@ -50,30 +66,194 @@ export default function Placeholder( {
+														navigationMenus.map(
+															( menu ) => {
+																return (
+																	<MenuItem
+																		onClick={ () => {
+																			setSelectedMenu(
+																				menu.id
+																			);
+																			onFinish(
+																				menu
+																			);
+																		} }
+																		onClose={
+																			onClose
+																		}
+																		key={
+																			menu.id
+																		}
+																	>
+																		{
+																			menu
+																				.title
+																				.rendered
+																		}
+																	</MenuItem>
```
This code gets very deeply nested. It's very hard to read a side-by-side diff. I think it'd be good to consider breaking the placeholder down into separate components.

Or maybe just extract the dropdown menu for selecting an existing menu to a separate component.

talldan commented on packages/block-library/src/navigation/edit/placeholder/index.js:
```diff
@@ -50,30 +66,194 @@ export default function Placeholder( {
+								</div>
+								{ hasMenus || canSwitchNavigationMenu ? (
+									<DropdownMenu
+										text={ __( 'Add existing menu' ) }
```
I'm not sure 'Add' is the right word now that the `wp_navigation` menus are in this dropdown, since the user is directly editing the menu. Maybe 'Select existing menu'.

tellthemachines committed:
Fix all pages button type logic.

talldan commented on packages/block-library/src/navigation/edit/placeholder/index.js:
```diff
@@ -50,30 +66,194 @@ export default function Placeholder( {
+											</>
+										) }
+									</DropdownMenu>
+								) : undefined }
```
TIL that React now lets you render `undefined`.

tellthemachines commented:
> I also still see those two blue buttons. Not sure exactly why.

I had neglected to update the logic that renders "Add all pages" as primary when the existing menus dropdown doesn't show 🤦  It should be fixed to take into account both classic and CPT menus now.

I've also addressed the remaining feedback.

talldan reviewed (commented):
Looks good so far, thanks for improving this 😄

tellthemachines committed:
Consolidate loading states and improve nesting.

tellthemachines commented:
Thanks for the review @talldan! I think I've addressed everything now.

talldan reviewed (approved):
Looks good, thanks again for tackling this!
